Models/SentenceDecoder.py

```diff
@@ -51,135 +53,138 @@ def __init__(self, vocab_size, hidden_size):
 
-        # h_bt = self.relu(e_t + self.Wb(h_b))
+        y_t = self.softmax(self.Woh(h_t))
 
-        # s_t = self.relu(self.Wd(h_ft + h_bt))
+        return y_t, h_t
 
-        b_v = self.Whi(CNN_last_layer)
-        h_t = self.relu(self.Whx(x) + self.Whh(h_t_prev)) # TODO
-        y_t = self.softmax(self.Woh(h_t))
 
-        return s_t
-    
 class MyRNN(nn.Module):
-    def __init__(self, obs_dim, hidden_size, output_dim):
+    def __init__(self, vocab_size, hidden_size, CNN_last_layer, device):
         """Initialize RNN."""
         super().__init__()
-        self.obs_dim = obs_dim
+        self.vocab_size = vocab_size
         self.hidden_size = hidden_size
+        self.cnn_last_layer = CNN_last_layer
 
-        self.rnn_cell = MyRNNCell(obs_dim, hidden_size)
+        self.rnn_cell = MyRNNCell(vocab_size, hidden_size, CNN_last_layer.shape[0])  # TODO: Double check if the last layer size is correct
+
+        self.embedding = nn.Embedding(vocab_size, hidden_size)
+        self.device = device
 
-    def forward(self, x):
-        """Compute forward pass on sequence x.
-        
-        Input sequence x has shape (B x L x D), where:
-        B is batch size, L is sequence length, and D is the number of features.
+
+    def forward(self, captions):
+        """
+        Write forward function for the RNN decoder for the Karpathy & Li Fei-Fei paper.
+
+        Returns the predicted word sequence and the hidden state sequence.
         """
-        batch_size, seq_len, n_feat = x.size()
-        
+        # batch_size x seq_len x hidden_size
+        # B is batch size, L is sequence length, and D is the number of features.
+        embeddings = self.embedding(captions)
+        embeddings = torch.cat((torch.zeros(embeddings.shape[0], 1, embeddings.shape[2]), embeddings), dim=1)  # The START vector is just a 0 vector here
+
+        batch_size, seq_len, n_feat = embeddings.size()
+
         # Stores outputs of RNN cell
-        output_arr = torch.zeros((batch_size, seq_len, self.output_dim))
+        output_arr = torch.zeros((batch_size, seq_len, self.vocab_size))
         hidden_arr = torch.zeros((batch_size, seq_len, self.hidden_size))
-        
-        # Send to GPU. This is a gotcha, make sure to send Tensors created
-        # in a model to the same device as input Tensors.
-        output_arr = output_arr.float().to(x.device)
-        hidden_arr = hidden_arr.float().to(x.device)
 
-        hidden = self.init_hidden(batch_size, x.device)
+        # Send to GPU. Send Tensors created in a model to the same device as input Tensors.
+        embeddings = embeddings.to(self.device)
+        output_arr = output_arr.float().to(self.device)
+        hidden_arr = hidden_arr.float().to(self.device)
+
+        # Initialize RNN hidden state: h_0 is the 0 vector
+        hidden = torch.zeros(batch_size, self.hidden_size, device=self.device)
 
         for i in range(seq_len):
+            # h_0 is the 0 vector
```

Comment:
are these comments useful?

Reply:
Ah I had them in to help understand the flow of things. Can remove if needed.

Reply:
yea maybe remove them if they are not going to be useful

Reply:
Ok will do.